Source/CombatExtended/Harmony/Harmony_GenRadial.cs

```diff
@@ -79,54 +79,92 @@ internal static void Patch()
-            int r = (int)radsq;
-            int count = radialPatternNumCells[r - 1];
-            /*
-              If we raise the max radius above about 200, binary search becomes faster.
-              Below 200, the match will reliably be found within 64 tries, where linear memory access dominates.
-            int max_count = RadialPatternNumCells[r];
-            int span;
-            */
+
+            int radiusSquaredInt = (int)(radius * radius);
+            // The answer is at least (25/8 * radius * radius - 11)
+            int lowerBound = ((radiusSquaredInt * 25) >> 3) - 11;
 #if BINSEARCH
-            while ((span = max_count - count) > 64)
+            // The upper bound is set to (101/32 * radius * radius)
+            //   some small radius actually goes above that
+            //   but the error is small (< 16) and linear search can fix it
+            int upperBound = (radiusSquaredInt * 101) >> 5;
+            if (upperBound > RadialPatternCount)
+                upperBound = RadialPatternCount;
+            // If the gap is large, perform binary search
+            while (upperBound - lowerBound > 88)
             {
-                int mid_count = count + span / 2;
-                float mid = radialPatternRadii[mid_count];
-                if (mid > radius)
-                {
-                    max_count = mid_count;
-                }
+                int mid = (lowerBound + upperBound) >> 1;
+                if (RadialPatternRadii[mid] <= radius)
+                    lowerBound = mid;
                 else
-                {
-                    count = mid_count;
-                }
+                    upperBound = mid;
             }
 #endif
-            float start = radialPatternRadii[count];
-            while (start <= radius)
-            {
-                start = radialPatternRadii[count++];
-            }
-            __result = count - 1;
+            // Linear search:
+            // When (int)radius + (int)diagonal is odd,  the answer is always 8n + 5
+            // When (int)radius + (int)diagonal is even, the answer is always 8n + 1
+            // So we can set the last three bits of lowerBound
+            // and check only every 8 numbers
+            lowerBound = (lowerBound & -7) | (HasOddCells(radius) ? 5 : 1);
+            while (RadialPatternRadii[lowerBound] <= radius)
+                lowerBound += 8;
+            __result = lowerBound;
             return false;
-
+        }
+        
+        // A helper function for Prefix_NumCellsInRadius()
+        // Evaluates if (int)r + (int)d is an even number (round toward zero)
+        // Where d = r * sqrt(1/2)
+        // When (int)radius + (int)diagonal is odd,  the answer is always 8n + 5
+        // When (int)radius + (int)diagonal is even, the answer is always 8n + 1
+        // Use IEEE-754 bit hacks
+        //[MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool HasOddCells(float r)
```

Comment:
This would also benefit from a pointer to external resources. At some point, we have to assume whomever ends up working on this next is reasonably capable, but pointing a novice in the right direction is good practice.  Otherwise some future dev may have to simply replace this with a lower performing version they can understand.